File: apps/api/src/middleware/clerk.ts
import type { NextFunction, Request, Response } from "express";

import { config } from "@/config/index.js";
import { um } from "@/lib/api-user-message.js";
import { sendJsonError } from "@/lib/http-json-response.js";
import { createLogger } from "@/utils/logger.js";
import { toLoggableError } from "@/utils/to-loggable-error.js";
import { verifyToken } from "@clerk/backend";

export async function clerkMiddleware(req: Request, res: Response, next: NextFunction) {
  const logger = createLogger("middleware:clerk");
  try {
    const authHeader = req.headers.authorization;
    if (!authHeader) {
      logger.warn("No authorization header provided");
      return sendJsonError(res, 401, "Unauthorized", um("UNAUTHORIZED", "unauthorized", "Unauthorized"));
    }
    const token = authHeader.replace("Bearer ", "");
    const payload = await verifyToken(token, {
      secretKey: config.clerkSecretKey,
    });
    logger.info("Clerk token verified for user: %s", payload.sub);

    req.auth = payload;
    next();
  } catch (error: unknown) {
    logger.error(toLoggableError(error), "Clerk token verification error");
    return sendJsonError(res, 401, "Unauthorized", um("UNAUTHORIZED", "unauthorized", "Unauthorized"));
  }
}

File: apps/api/src-go/internal/middleware/clerk.go
package middleware

import (
	"strings"

	"github.com/labstack/echo/v4"

	"linky-api/src-go/internal/infra/clerkx"
	"linky-api/src-go/internal/httpx"
	"linky-api/src-go/internal/logger"
)

var clerkLog = logger.New("middleware:clerk")

func Clerk() echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			authHeader := c.Request().Header.Get("Authorization")
			if authHeader == "" {
				clerkLog.Warn().Msg("No authorization header provided")
				return httpx.Unauthorized(c)
			}
			token := strings.TrimPrefix(authHeader, "Bearer ")
			payload, err := clerkx.VerifyToken(c.Request().Context(), token)
			if err != nil {
				clerkLog.Error().Err(err).Msg("Clerk token verification error")
				return httpx.Unauthorized(c)
			}
			httpx.SetAuth(c, &httpx.AuthClaims{Sub: payload.Sub, Raw: payload.Raw})
			return next(c)
		}
	}
}
